### PR TITLE
Optimize React on `next build`

### LINKED ISSUE #937: Remove propTypes in production build
Implement something like this:
https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types

### PR DESCRIPTION
**Pull request by timneutkens**
Fixes #937

Implements all but one plugin from https://github.com/thejameskyle/babel-react-optimize. https://github.com/thejameskyle/babel-react-optimize#transform-react-inline-elements conflicts with styled-jsx at the moment. @giuseppeg is looking into that right now.

### CODE REVIEW

**Comment by timneutkens** [comment]
cc @nkzawa @rauchg

**Comment by nkzawa** [comment]
👍  awesome

**Comment by timneutkens** [comment]
👍

**Comment by frol** [comment]
FYI, `transform-react-pure-class-to-function` is not safe because Refs don't work on stateless function components: https://github.com/facebook/react/issues/4936

I am not sure how babel deals with it, but it may break some code: https://github.com/zeit/next.js/blob/master/examples/with-react-md/pages/index.js#L34-L37

**Comment by timneutkens** [comment]
This is why I love OSS, thanks for your feedback @frol ❤️, removing it.

File: package.json [hunk 1]
```diff
@@ -52,6 +52,9 @@
     "babel-plugin-transform-class-properties": "6.22.0",
     "babel-plugin-transform-es2015-modules-commonjs": "6.22.0",
     "babel-plugin-transform-object-rest-spread": "6.22.0",
+    "babel-plugin-transform-react-constant-elements": "^6.22.0",
+    "babel-plugin-transform-react-pure-class-to-function": "^1.0.1",
+    "babel-plugin-transform-react-remove-prop-types": "^0.3.2",
```
Comment (oliviertassinari) [inline package.json]: ❤️

Comment (timneutkens) [inline package.json]: 👍

**Pull request merged**